File: python/Backtracking/77.combinations.py
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res: List[List[int]] = []

        def backtrack(first: int, curr: List[int]):
            if len(curr) == k:
                res.append(curr)
            for i in range(first, n + 1):
                backtrack(i + 1, curr + [i])

        backtrack(1, [])
        return res


a = Solution()
a.combine(2, 4)

File: TypeScript/Backtracking/77.combinations.ts
function combine(n: number, k: number): number[][] {
  let res: number[][] = []
  function backtrack(curr: number[], start: number) {
    if (curr.length == k) {
      res.push([...curr])
      return
    }
    for (let i = start; i <= n; ++i) {
      backtrack([...curr, i], i + 1)
    }
  }
  backtrack([], 1)
  return res
}
